internal/controllers/topology/cluster/desired_state.go

```diff
@@ -758,16 +758,23 @@ func computeMachineDeployment(_ context.Context, s *scope.Scope, desiredControlP
 // computeMachineDeploymentVersion calculates the version of the desired machine deployment.
 // The version is calculated using the state of the current machine deployments,
 // the current control plane and the version defined in the topology.
-// Nb: No MachineDeployment upgrades will be triggered while any MachineDeployment is in the middle
-// of an upgrade. Even if the number of MachineDeployments that are being upgraded is less
-// than the number of allowed concurrent upgrades.
-func computeMachineDeploymentVersion(s *scope.Scope, machineDeploymentTopology clusterv1.MachineDeploymentTopology, desiredControlPlaneState *scope.ControlPlaneState, currentMDState *scope.MachineDeploymentState) (string, error) {
+func computeMachineDeploymentVersion(s *scope.Scope, desiredMDName string, machineDeploymentTopology clusterv1.MachineDeploymentTopology, currentMDState *scope.MachineDeploymentState) string {
 	desiredVersion := s.Blueprint.Topology.Version
-	// If creating a new machine deployment, we can pick up the desired version
-	// Note: We are not blocking the creation of new machine deployments when
-	// the control plane or any of the machine deployments are upgrading/scaling.
+	// If creating a new machine deployment mark it as pending if the control plane is not
+	// yet stable. Creating a new MD while the control plane is upgrading is unsafe.
 	if currentMDState == nil || currentMDState.Object == nil {
-		return desiredVersion, nil
+		if s.UpgradeTracker.ControlPlane.IsStartingUpgrade ||
+			s.UpgradeTracker.ControlPlane.IsUpgrading ||
+			s.UpgradeTracker.ControlPlane.IsScaling ||
+			s.UpgradeTracker.ControlPlane.IsProvisioning {
+			// TODO(ykakarap): Here we are marking as pending upgrade so that the reconciliation of this MD does not
+			// occur and therefore not create the new MD. Find a better name.
```

Comment:
I think it would make sense to differentiate between upgrade pending and creation pending

This way we get the exact right information in the UpgradeTracker and can also write the conditions accordingly